tests/matrixstore/test_service.py

```diff
@@ -1,130 +1,144 @@
-        user_service=Mock(),
-    )
-    id = service.create(dto)
-
-    # Verify
-    assert id == "my-id"
-    repo.save.assert_called_once_with(matrix)
-    repo_content.save.assert_called_once_with(dto)
-
-
-def test_get():
-    # Init Mock
-    content = Mock()
-    content.get.return_value = MatrixContent(
-        data=[[1, 2]],
-        index=[1],
-        columns=["a", "b"],
-    )
-
-    repo = Mock()
-    repo.get.return_value = Matrix(
-        id="my-id",
-        width=2,
-        height=1,
-        created_at=datetime.datetime.fromtimestamp(42),
-    )
-
-    repo_meta = Mock()
-
-    # Expected
-    exp = MatrixDTO.construct(
-        id="my-id",
-        created_at=42,
-        width=2,
-        height=1,
-        data=[[1.0, 2.0]],
-        index=[1],
-        columns=["a", "b"],
-    )
-
-    # Test
-    service = MatrixService(
-        repo, repo_meta, content, Mock(), Mock(), Mock(), Mock()
-    )
-    res = service.get("my-id")
-    assert exp == res
-
-
-def test_delete():
-    content = Mock()
-    repo = Mock()
-    repo_meta = Mock()
-
-    service = MatrixService(
-        repo, repo_meta, content, Mock(), Mock(), Mock(), Mock()
-    )
-    service.delete("my-id")
-    content.delete.assert_called_once_with("my-id")
-    repo.delete.assert_called_once_with("my-id")
+class TestMatrixService:
+    def test_create__nominal_case(self, matrix_service: MatrixService):
+        """Creates a new matrix object with the specified data."""
+        # when a matrix is created (inserted) in the service
+        data = [[1, 2, 3], [4, 5, 6]]
+        matrix_hash = matrix_service.create(data)
+
+        # A "real" hash value is calculated
+        assert matrix_hash, "ID can't be empty"
+
+        # The matrix is saved in the content repository as a TSV file
+        bucket_dir = matrix_service.matrix_content_repository.bucket_dir
+        content_path = bucket_dir.joinpath(f"{matrix_hash}.tsv")
+        array = np.loadtxt(content_path)
+        assert array.all() == np.array(data).all()
+
+        # A matrix object is stored in the database
+        with db():
+            obj = matrix_service.repo.get(matrix_hash)
+        assert obj is not None, f"Missing Matrix object {matrix_hash}"
+        assert obj.width == len(data[0])
+        assert obj.height == len(data)
+        now = datetime.datetime.utcnow()
+        assert now - datetime.timedelta(seconds=1) <= obj.created_at <= now
+
+    def test_create__side_effect(self, matrix_service: MatrixService):
+        """Creates a new matrix object with the specified data, but fail during saving."""
+        # if the matrix can't be created in the service
+        matrix_repo = matrix_service.repo
+        matrix_repo.save = Mock(side_effect=Exception("database error"))
+        with pytest.raises(Exception, match="database error"):
+            data = [[1, 2, 3], [4, 5, 6]]
+            matrix_service.create(data)
+
+        # The matrix is not saved in the content repository
```

Comment:
here the test is wrong : 
the file should still be here if the matrix existed before.

In the end, I think it's better to leave the garbage collector handle this, as it was done before.

Reply:
corrected